libs/@local/graph/migrations-macros/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+extern crate alloc;
+
+use alloc::collections::{BTreeMap, btree_map::Entry};
+use core::{error::Error, num::ParseIntError};
+use std::{
+    fs, io,
+    path::{Path, PathBuf},
+    sync::OnceLock,
+};
+
+use convert_case::{Case, Casing as _};
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use regex::Regex;
+use sha2::{Digest as _, Sha256};
+use syn::{LitStr, parse_macro_input};
+use walkdir::{DirEntry, WalkDir};
+
+pub(crate) fn crate_root() -> PathBuf {
+    PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").expect("Could not read crate root"))
+}
+
+#[derive(Debug, Clone)]
+struct MigrationFile {
+    module_path: PathBuf,
+    name: String,
+    number: u32,
+    size: usize,
+    digest: [u8; 32],
+}
+
+fn read_directory_content(path: impl AsRef<Path>) -> io::Result<String> {
```

Comment:
in this case, merely renaming a file wouldn't trigger a change in digest.

You also don't need to have a single buffer that you return. You can also just:

```rs
fn hash_directory_contents(path: impl AsRef<Path>, digest: &mut impl Digest) -> io::Result<String> {
	for path in ... {
		digest.update(entry.file_name());
		digest.update(&...);
	}
}

// in the caller you can then just call `digest.finalize`
```

I also don't know if `fs::read_to_string` is correct here, that means it'd require any dependent file to be valid utf-8, so binary blobs couldn't be used.

Reply:
Good point, I updated the digest generation.